### PR TITLE
Remove conflicts on rector/rector

### PR DESCRIPTION
**Pull request by mglaman**
## Description
Removes conflicts against rector/rector.

Uses the Composer installed.json from each tag to find references for require-dev. Documented this in README.md

We cannot use 0.12.18 as it has a hack in its `bootstrap.php` around `TestCase` which autoloads too early and breaks Drupal's workaround for `TestCase` compatibility. It was removed in 0.12.19.

## To Test
Normal testing

## Drupal.org issue
https://www.drupal.org/project/rector/issues/3271102

Let's have this PR just address that issue. We can handle 0.12.20+ in another.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -10,11 +10,11 @@
         "ast"
     ],
     "require": {
-        "rector/rector": "~0.12",
+        "rector/rector": "~0.12.19",
         "webflo/drupal-finder": "^1.2"
     },
     "conflict": {
-        "rector/rector": ">0.12.17"
+        "rector/rector": ">0.12.19"
```
Comment (mglaman) [inline composer.json:17]: Right now, 0.12.19 is the "sweet spot" for drupal-rector working.

File: composer.json [hunk 2]
```diff
@@ -10,11 +10,11 @@
         "ast"
     ],
     "require": {
-        "rector/rector": "~0.12",
+        "rector/rector": "~0.12.19",
```
Comment (mglaman) [inline composer.json:13]: We're broken against 0.12.18 and it's `bootstrap.php` code which autoloads TestCase too early and Drupal cannot rewrite it.

File: config/drupal-phpunit-bootstrap-file.php [hunk 3]
```diff
@@ -169,24 +168,10 @@ function drupal_phpunit_populate_class_loader($drupalRoot, $vendorRoot) {
...
-    // @link https://git.drupalcode.org/project/drupal/-/blob/9.3.x/core/tests/Drupal/TestTools/PhpUnitCompatibility/PhpUnit8/ClassWriter.php
-    if ($major >= 8) {
-        $loader = require $drupalVendorRoot . '/autoload.php';
-        \DrupalRector\Utility\Drupal9\ClassWriter::mutateTestBase($loader);
+$autoloader = require $drupalVendorRoot . '/autoload.php';
+if ($autoloader instanceof \Composer\Autoload\ClassLoader) {
+    if (interface_exists(\PHPUnit\Framework\Test::class)
+        && class_exists('Drupal\TestTools\PhpUnitCompatibility\PhpUnit8\ClassWriter')) {
+        \Drupal\TestTools\PhpUnitCompatibility\PhpUnit8\ClassWriter::mutateTestBase($autoloader);
     }
```
Comment (mglaman) [inline config/drupal-phpunit-bootstrap-file.php:176]: Updated this to match what phpstan-drupal does, and now we don't need to have a forked copy of the class writer.

**Comment by agentolivia** [comment]
What does "Normal testing" mean?

**Comment by mglaman** [comment]
😅 ah yeah. It's a normal test run, but I guess it needs a description on how to install this branch into your Drupal site.

I'll update later with that info.

**Comment by agentolivia** [comment]
I would love to help, but I'm new to this project and need at least a step 1.

**Comment by bbrala** [comment]
I've run this branch succesfully in the project_update enviroment as part of d10 readiness. Only noticed a new extension might be preferrred.

`Notice: RectorPrefix20220323\Nette\Utils\Strings::toAscii(): it is recommended to enable PHP extensions 'intl'. in /var/lib/drupalci/workspace/drupal-checkouts/drupal9/vendor/rector/rector/vendor/nette/utils/src/Utils/Strings.php on line 124` 

Attached the output, this also includes the output of rector.

[Uploading phpstan-results.zip…]()

**Comment by bbrala** [comment]
Hmm, attachment failed
[phpstan-results.zip](https://github.com/palantirnet/drupal-rector/files/8748077/phpstan-results.zip)

**Comment by mglaman** [comment]
@agentolivia sorry I haven't had time to give details on how to test. I plan on writing a general blog for it and also contributing it to the README.

It requires setting the vcs repo for Composer, and then requiring dev-main#HASHOFLASTPRCOMMIT

**Pull request merged**